scispacy/candidate_generation.py

```diff
@@ -0,0 +1,307 @@
+                mention = self.ann_concept_aliases_list[neighbor_index]
+                concepts_for_mention = self.mention_to_concept[mention]
+                for concept_id in concepts_for_mention:
+                    concept_to_mentions[concept_id].append(mention)
+                    concept_to_distances[concept_id].append(distance)
+
+            mention_candidates = [MentionCandidate(concept, mentions, concept_to_distances[concept])
+                                  for concept, mentions in concept_to_mentions.items()]
+
+            neighbors_by_concept_ids.append(mention_candidates)
+
+        return neighbors_by_concept_ids
+
+
+def create_tfidf_ann_index(out_path: str, umls: List = None) -> Tuple[List[str], TfidfVectorizer, FloatIndex]:
+    """
+    Build tfidf vectorizer and ann index.
+
+    Warning: Running this function on the whole of UMLS requires ~ 200GB of RAM ...
+    TODO: Make this not take 200GB of RAM.
+
+    Parameters
+    ----------
+    out_path: str, required.
+        The path where the various model pieces will be saved.
+    umls : List, optional.
+        The list of umls kb items to generate the index and vectors for.
+
+    """
+    tfidf_vectorizer_path = f'{out_path}/tfidf_vectorizer.joblib'
+    ann_index_path = f'{out_path}/nmslib_index.bin'
+    tfidf_vectors_path = f'{out_path}/tfidf_vectors_sparse.npz'
+    uml_concept_aliases_path = f'{out_path}/concept_aliases.json'
+
+    umls = umls or json.load(open(cached_path(DEFAULT_PATHS["umls_path"])))
+    # We need to keep around a map from text to possible canonical ids that they map to.
+    text_to_concept_id: Dict[str, Set[str]] = defaultdict(set)
+
+    for concept in umls:
+        for alias in set(concept["aliases"]).union({concept["canonical_name"]}):
+            text_to_concept_id[alias].add(concept["concept_id"])
+
+    # nmslib hyperparameters (very important)
+    # guide: https://github.com/nmslib/nmslib/blob/master/python_bindings/parameters.md
+    # Default values resulted in very low recall.
+
+    # set to the maximum recommended value. Improves recall at the expense of longer indexing time.
+    # TODO: This variable name is so hot because I don't actually know what this parameter does.
+    m_parameter = 100
+    # `C` for Construction. Set to the maximum recommended value
+    # Improves recall at the expense of longer indexing time
+    construction = 2000
+    num_threads = 60  # set based on the machine
+    index_params = {'M': m_parameter, 'indexThreadQty': num_threads, 'efConstruction': construction, 'post' : 0}
+
+    print(f'No tfidf vectorizer on {tfidf_vectorizer_path} or ann index on {ann_index_path}')
+    umls_concept_aliases = list(text_to_concept_id.keys())
+
+    umls_concept_aliases = numpy.array(umls_concept_aliases)
+
+    print(f'Fitting tfidf vectorizer on {len(umls_concept_aliases)} aliases')
+    tfidf_vectorizer = TfidfVectorizer(analyzer='char_wb', ngram_range=(3, 3), min_df=10, dtype=numpy.float32)
```

Comment:
I know we talked about this a bunch, but I can't remember what the outcome was. Can we change this to float16?